lib/ansible/plugins/test/core.py

```diff
@@ -149,7 +149,10 @@ def regex(value='', pattern='', ignorecase=False, multiline=False, match_type='s
     if multiline:
         flags |= re.M
     _re = re.compile(pattern, flags=flags)
-    return bool(getattr(_re, match_type, 'search')(value))
+    try:
+        return bool(getattr(_re, match_type)(value))
+    except (TypeError, AttributeError) as exc:
+        raise errors.AnsibleTemplatePluginError("The regex test expects a valid match_type such as 'search' or 'match'.", obj=match_type) from exc
```

Comment:
Looking at this a bit more, I think we should restrict `match_type` to a known list of methods instead of allowing arbitrary methods to be invoked. We already document `search` and `match`, but there is at least one more that makes sense, which is `fullmatch`. Those methods should be included, and added to the documented choices.

Also, regarding the error message, mentioning the the plugin name and type is now redundant, so whatever error message we use can omit those.